client/src/main/java/com/hivemq/client/internal/mqtt/codec/encoder/MqttEncoder.java

```diff
@@ -55,8 +64,51 @@
         context = new MqttEncoderContext(ByteBufAllocator.DEFAULT);
     }
 
+    public void onConnected(final @NotNull MqttClientConnectionConfig connectionConfig,
+                            final @NotNull MqttClientConfig clientConfig) {
+      context.setMaximumPacketSize(connectionConfig.getSendMaximumPacketSize());
+      this.priorities = clientConfig.getTopicPriorities();
+    }
+
     public void onConnected(final @NotNull MqttClientConnectionConfig connectionConfig) {
-        context.setMaximumPacketSize(connectionConfig.getSendMaximumPacketSize());
+      context.setMaximumPacketSize(connectionConfig.getSendMaximumPacketSize());
+    }
+
+    public void setTOS(final @NotNull ChannelHandlerContext ctx,
+                       final @NotNull Object msg) {
+
+      MqttTopic topic = (msg instanceof MqttPublish) 
+                      ? ((MqttPublish) msg).getTopic()
+                      : (msg instanceof MqttStatefulPublish)
+                      ? ((MqttStatefulPublish) msg).stateless().getTopic()
+                      : null;
+
+      if (topic == null) return;
+
+      if (!(ctx.channel().config() instanceof SocketChannelConfig)) return;
+      SocketChannelConfig config = ((SocketChannelConfig) ctx.channel().config());
+
+      PriorityClass priority = 
+        priorities.stream()
+                  .filter(p -> p.getTopicFilter().matches(topic.filter()))
+                  .max(Comparator.comparingInt(a -> a.getTopicFilter().getLevels().size()))
+                  .map(a -> a.getPriorityClass())
+                  .orElseGet(() -> PriorityClass.ROUTINE);
+                  
+      int prevTOS = config.getTrafficClass();
+      int newTOS  = 0;
+      
+      switch (priority) {
```

Comment:
I think this doesn't work quite like you expect. 

The socket used for communication from client to broker is persistent. This means that if you set a bit in the header for a specific message all following messages will also have that bit set. Which means that if i first send on with **PRIORITY** nad then one with **IMMEDIEATE** the header priority will be 11 because of the bit set in the first message. You need to somehow reset every time you want to send  a new message.

If you try to use the code bellow in the test app you might see what I mean. 

```
    val client = Mqtt5Client.builder()
        .identifier(UUID.randomUUID().toString())
        .serverHost("165.22.126.152")
        .addTopicPriority(TopicPriority(MqttTopicFilter.of("routine/#"), PriorityClass.ROUTINE))
        .addTopicPriority(TopicPriority(MqttTopicFilter.of("priority/#"), PriorityClass.PRIORITY))
        .addTopicPriority(TopicPriority(MqttTopicFilter.of("immediate/#"), PriorityClass.IMMEDIATE))
        .addTopicPriority(TopicPriority(MqttTopicFilter.of("flash/#"), PriorityClass.FLASH))
        .buildBlocking()

    client.connect()
    client.publishWith().topic("routine/1").payload("Hei".toByteArray()).qos(MqttQos.AT_MOST_ONCE).send()
    client.publishWith().topic("priority/1").payload("Hei".toByteArray()).qos(MqttQos.AT_MOST_ONCE).send()
    client.publishWith().topic("immediate/1").payload("Hei".toByteArray()).qos(MqttQos.AT_MOST_ONCE).send()
    client.publishWith().topic("flash/1").payload("Hei".toByteArray()).qos(MqttQos.AT_MOST_ONCE).send()
```

Reply:
Oi, fan.. hadde det, men kutta litt aggresivt naar jeg rydda opp koden